### ISSUE TITLE
duplicate card bug can also result in running out of cards

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#136]
First there were reports of people getting duplicate cards, which hasn't happened before, but I'll make a separate ticket for that. Then we had a crash. Turns out the crash was related to the duplicate cards (as I explain in the comments).
```
2020-04-12 19:12:34,512 INFO  [actix_server::builder] Starting "actix-web-service-0.0.0.0:8081" service on 0.0.0.0:8081
request: ClosedConnection(TableSessionId(726962610757016154))
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', mb2/src/table.rs:1653:40
stack backtrace:
   0:     0x55c949130e04 - backtrace::backtrace::libunwind::trace::he392e7f3190d39cb
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x55c949130e04 - backtrace::backtrace::trace_unsynchronized::h6dce47f25b374263
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x55c949130e04 - std::sys_common::backtrace::_print_fmt::heee0ad6502408912
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x55c949130e04 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h74e69a0924964693
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x55c9491596ac - core::fmt::write::h8a9d34ce2fb2b006
                               at src/libcore/fmt/mod.rs:1069
   5:     0x55c949129c83 - std::io::Write::write_fmt::h0f9fe59759aa4586
                               at src/libstd/io/mod.rs:1504
   6:     0x55c9491335e5 - std::sys_common::backtrace::_print::h3cf02f93364030ac
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x55c9491335e5 - std::sys_common::backtrace::print::hf647a928a22b922a
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x55c9491335e5 - std::panicking::default_hook::{{closure}}::he87f3d294472900e
                               at src/libstd/panicking.rs:198
   9:     0x55c949133322 - std::panicking::default_hook::h5a38802aba5b8783
                               at src/libstd/panicking.rs:218
  10:     0x55c949133c42 - std::panicking::rust_panic_with_hook::h67ec5890687fb40f
                               at src/libstd/panicking.rs:511
  11:     0x55c94913382b - rust_begin_unwind
                               at src/libstd/panicking.rs:419
  12:     0x55c949157e21 - core::panicking::panic_fmt::hc45bcde77dfd550f
                               at src/libcore/panicking.rs:111
  13:     0x55c949157d6d - core::panicking::panic::h9a3361bbdf10a542
                               at src/libcore/panicking.rs:54
  14:     0x55c948ec99fa - core::option::Option<T>::unwrap::hb1b704f2bf53839a
                               at /rustc/e82734e56b2a50d38e0937d08f559d15dbe8e46b/src/libcore/macros/mod.rs:34
  15:     0x55c948ec99fa - mb2::table::Table::discard_or_draw::h9703c03832750253
                               at mb2/src/table.rs:1653
  16:     0x55c948ec893c - mb2::table::Table::do_advanced::h3dccf6bd60e25f7c
                               at mb2/src/table.rs:1562
```

**Comment by ctm** [comment]
This first crash is that we ran out of cards. This is probably due to how I implemented shuffling the muck., which is almost undoubtedly wrong, seeing how we got #137.

The second crash was due to us playing a tournament with 8 players that only holds 6 players.

**Comment by ctm** [comment]
I believe the way we ran out of cards is due to the duplication of cards (#137), because we have this code:
```
        let shuffled_muck_message = if is_drawing && deck.len() < n {
            deck.shuffle_in(&mut self.muck);
...
        }
...
        {
...        
            for card in player.cards.iter_mut() {
                if cards.contains(card) {
                    card.replace_inner(new_cards.next().unwrap());
...
```
so in that for loop, if someone asks to discard a duplicated card, then we'll need **2** cards to replace it, but above we only decided if we needed to shuffle the muck based on the number of cards that were being discarded. So, let's say there's exactly one card left in the deck, but you have two sixes of spades. You think you want to get rid of one and so you say `d 6s`. The code then decides there's enough cards left in the deck, but as it's going through your hand replacing cards it needs a new card for **each** `6s`, so we run out.
```
